onnx/defs/nn/defs.cc

```diff
@@ -3443,6 +3445,8 @@ ONNX_OPERATOR_SET_SCHEMA(
             "Shape must be broadcastable to "
             "4D tensor with shape `(batch_size, q_num_heads, q_sequence_length, total_sequence_length)` "
             "where `total_sequence_length = past_sequence_length + kv_sequence_length.` "
+            "The kv_sequence dimension can be smaller than `total_sequence_length` in the case where `K` and `V` are the entire cache tensor "
```

Comment:
I think it would be better to update the previous two lines also, using a different name than total_sequence_length. But do we really need the generalization? It is likely to confuse people (since we allow smaller, but not larger mask, we allow this in last-dimension, but not others like q_sequence_length). In practice, if users are going to use nonpad_kv_seq_len, they may not use this mask, right?

Reply:
Ok. I suggest changing `total_sequence_length` in previous two lines to some other word: may be `attn_sequence_length` or something like that. Then, the documentation can say that `attn_sequence_length` must be <= total_sequence_length, and that it will be padded to total_sequence_length with minus infinity values if it is smaller.

Reply:
Actually I think we can just say this
```
            "Shape must be broadcastable to `(batch_size, q_num_heads, q_sequence_length, total_sequence_length)` "
            "where `total_sequence_length = past_sequence_length + kv_sequence_length.` "
            "The last dimension can also be shorter than `total_sequence_length` and will be padded to `total_sequence_length` with negative infinity, "
            "but it still needs to be at least as long as the maximum value of `nonpad_kv_seqlen` if provided. "
```
This should cover all three cases: 
1. no kv cache at all 
2. kv cache concat happens inside Attention op with past/present_key/value I/O
3. K/V inputs are entire cache buffer

Reply:
That's fine. I still have one quibble about the last line "but it still needs to be at least as long as the maximum ..." ... I don't think we are checking for this condition in the shape-inference, are we? If we are not checking for this condition, it would be simpler to drop this condition here ... because padding to the required size can be done no matter what its length is, as long as it is smaller. And if this works for the user, then that's fine.

Reply:
Yea that's right, we are not checking this condition in shape inference. Let me drop this then.